File: plugins/svn4idea/src/org/jetbrains/idea/svn/IdeaSVNConfigFile.java
// Copyright 2000-2017 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.jetbrains.idea.svn;

import com.intellij.openapi.util.Couple;
import com.intellij.openapi.util.text.StringUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.idea.svn.api.Url;
import org.jetbrains.idea.svn.commandLine.SvnBindException;
import org.jetbrains.idea.svn.config.DefaultProxyGroup;
import org.jetbrains.idea.svn.config.ProxyGroup;
import org.tmatesoft.svn.core.internal.wc.DefaultSVNOptions;
import org.tmatesoft.svn.core.internal.wc.SVNConfigFile;

import java.io.File;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

import static com.intellij.util.containers.ContainerUtil.union;
import static org.jetbrains.idea.svn.SvnUtil.createUrl;

public class IdeaSVNConfigFile {

  public final static String SERVERS_FILE_NAME = "servers";
  public final static String CONFIG_FILE_NAME = "config";

  public static final String DEFAULT_GROUP_NAME = "global";
  public static final String GROUPS_GROUP_NAME = "groups";

  private final Map<String, String> myPatternsMap;
  private final long myLatestUpdate;
  private final File myFile;
  private Map myDefaultProperties;

  private final SVNConfigFile mySVNConfigFile;


  public IdeaSVNConfigFile(final File file) {
    mySVNConfigFile = new SVNConfigFile(file);
    myFile = file;
    myLatestUpdate = -1;
    myPatternsMap = new HashMap<>();
  }

  @NotNull
  public static String getNewGroupName(@NotNull String host, @NotNull IdeaSVNConfigFile configFile) {
    String groupName = host;
    final Map<String, ProxyGroup> groups = configFile.getAllGroups();
    while (StringUtil.isEmptyOrSpaces(groupName) || groups.containsKey(groupName)) {
      groupName += "1";
    }
    return groupName;
  }

  public void updateGroups() {
    if (myLatestUpdate != myFile.lastModified()) {
      myPatternsMap.clear();
      myPatternsMap.putAll(mySVNConfigFile.getProperties(GROUPS_GROUP_NAME));

      myDefaultProperties = mySVNConfigFile.getProperties(DEFAULT_GROUP_NAME);
    }
  }

  public Map<String, ProxyGroup> getAllGroups() {
    final Map<String, ProxyGroup> result = new HashMap<>(myPatternsMap.size());
    for (Map.Entry<String, String> entry : myPatternsMap.entrySet()) {
      final String groupName = entry.getKey();
      result.put(groupName, new ProxyGroup(groupName, entry.getValue(), mySVNConfigFile.getProperties(groupName)));
    }
    return result;
  }

  public DefaultProxyGroup getDefaultGroup() {
    return new DefaultProxyGroup(myDefaultProperties);
  }

  @Nullable
  public String getValue(String groupName, String propertyName) {
    return mySVNConfigFile.getPropertyValue(groupName, propertyName);
  }

  @NotNull
  public Map<String, String> getValues(String groupName) {
    return mySVNConfigFile.getProperties(groupName);
  }

  public void setValue(final String groupName, final String propertyName, final String value) {
    mySVNConfigFile.setPropertyValue(groupName, propertyName, value, true);
  }

  public void deleteGroup(final String name) {
    // remove all properties
    final Map<String, String> properties = mySVNConfigFile.getProperties(name);
    for (String propertyName : properties.keySet()) {
      mySVNConfigFile.setPropertyValue(name, propertyName, null, false);
    }
    if (DEFAULT_GROUP_NAME.equals(name)) {
      myDefaultProperties.clear();
    }
    // remove group from groups
    mySVNConfigFile.setPropertyValue(GROUPS_GROUP_NAME, name, null, false);
    mySVNConfigFile.deleteGroup(name, false);
  }

  public void addGroup(final String name, final String patterns, final Map<String, String> properties) {
    mySVNConfigFile.setPropertyValue(GROUPS_GROUP_NAME, name, patterns, false);
    addProperties(name, properties);
  }

  private void addProperties(final String groupName, final Map<String, String> properties) {
    for (Map.Entry<String, String> entry : properties.entrySet()) {
      mySVNConfigFile.setPropertyValue(groupName, entry.getKey(), entry.getValue(), false);
    }
  }

  public void modifyGroup(final String name, final String patterns, final Collection<String> delete, final Map<String, String> addOrModify,
                          final boolean isDefault) {
    if (!isDefault) {
      mySVNConfigFile.setPropertyValue(GROUPS_GROUP_NAME, name, patterns, false);
    }
    final Map<String, String> deletedPrepared = new HashMap<>(delete.size());
    for (String property : delete) {
      deletedPrepared.put(property, null);
    }
    addProperties(name, deletedPrepared);
    addProperties(name, addOrModify);
  }

  public void save() {
    mySVNConfigFile.save();
  }

  public static String getPropertyIdea(String host, @NotNull Couple<IdeaSVNConfigFile> serversFile, final String name) {
    String groupName = getGroupName(getValues(serversFile, GROUPS_GROUP_NAME), host);
    if (groupName != null) {
      Map hostProps = getValues(serversFile, groupName);
      final String value = (String)hostProps.get(name);
      if (value != null) {
        return value;
      }
    }
    Map globalProps = getValues(serversFile, "global");
    return (String)globalProps.get(name);
  }

  public static boolean checkHostGroup(final String url, final String patterns, final String exceptions) {
    final Url svnurl;
    try {
      svnurl = createUrl(url);
    }
    catch (SvnBindException e) {
      return false;
    }

    final String host = svnurl.getHost();
    return matches(patterns, host) && (!matches(exceptions, host));
  }

  private static boolean matches(final String pattern, final String host) {
    final StringTokenizer tokenizer = new StringTokenizer(pattern, ",");
    while (tokenizer.hasMoreTokens()) {
      String token = tokenizer.nextToken();
      if (DefaultSVNOptions.matches(token, host)) {
        return true;
      }
    }
    return false;
  }

  @Nullable
  public static String getGroupForHost(final String host, final IdeaSVNConfigFile serversFile) {
    final Map<String, ProxyGroup> groups = serversFile.getAllGroups();
    for (Map.Entry<String, ProxyGroup> entry : groups.entrySet()) {
      if (matches(entry.getValue().getPatterns(), host)) return entry.getKey();
    }
    return null;
  }

  // taken from default manager as is
  private static String getGroupName(Map groups, String host) {
    for (Object o : groups.keySet()) {
      final String name = (String)o;
      final String pattern = (String)groups.get(name);
      if (matches(pattern, host)) return name;
    }
    return null;
  }

  // default = yes
  public static boolean isTurned(final String value) {
    return value == null || "yes".equalsIgnoreCase(value) || "on".equalsIgnoreCase(value) || "true".equalsIgnoreCase(value);
  }

  @Nullable
  public static String getValue(@NotNull Couple<IdeaSVNConfigFile> files, String groupName, String propertyName) {
    String result = files.second.getValue(groupName, propertyName);
    return result != null ? result : files.first.getValue(groupName, propertyName);
  }

  @NotNull
  public static Map<String, String> getValues(@NotNull Couple<IdeaSVNConfigFile> files, String groupName) {
    return union(files.first.getValues(groupName), files.second.getValues(groupName));
  }
}

File: plugins/svn4idea/src/org/jetbrains/idea/svn/IdeaSVNConfigFile.kt
// Copyright 2000-2017 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.jetbrains.idea.svn

import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.util.Couple
import com.intellij.openapi.util.io.FileSystemUtil.lastModified
import com.intellij.openapi.util.text.StringUtil
import com.intellij.util.containers.ContainerUtil.union
import org.apache.oro.text.GlobCompiler
import org.apache.oro.text.regex.MalformedPatternException
import org.apache.oro.text.regex.Perl5Matcher
import org.ini4j.Config
import org.ini4j.Ini
import org.ini4j.spi.IniBuilder
import org.ini4j.spi.IniFormatter
import org.jetbrains.idea.svn.SvnUtil.createUrl
import org.jetbrains.idea.svn.api.Url
import org.jetbrains.idea.svn.commandLine.SvnBindException
import org.jetbrains.idea.svn.config.DefaultProxyGroup
import org.jetbrains.idea.svn.config.ProxyGroup
import java.io.IOException
import java.io.PrintWriter
import java.io.Writer
import java.nio.file.Path
import java.util.*

private val LOG = logger<IdeaSVNConfigFile>()

class IdeaSVNConfigFile(private val myPath: Path) {

  private val myPatternsMap = mutableMapOf<String, String?>()
  private var myLatestUpdate = -1L
  private val myDefaultProperties = mutableMapOf<String, String?>()
  private val _configFile = MyIni().apply {
    config.isTree = false
    config.isEmptySection = true
  }
  private val configFile: Ini
    get() {
      updateGroups()
      return _configFile
    }

  val allGroups: Map<String, ProxyGroup>
    get() {
      val result = HashMap<String, ProxyGroup>(myPatternsMap.size)
      for ((groupName, value) in myPatternsMap) {
        result.put(groupName, ProxyGroup(groupName, value, getValues(groupName)))
      }
      return result
    }

  val defaultGroup get() = DefaultProxyGroup(myDefaultProperties)

  @JvmOverloads
  fun updateGroups(force: Boolean = false) {
    val lastModified = lastModified(myPath.toFile())

    if (force || myLatestUpdate != lastModified) {
      _configFile.clear()
      try {
        _configFile.load(myPath.toFile())
      }
      catch (e: IOException) {
        LOG.info("Could not load $myPath", e)
      }
      myLatestUpdate = lastModified

      myPatternsMap.clear()
      myPatternsMap.putAll(getValues(GROUPS_GROUP_NAME))

      myDefaultProperties.clear()
      myDefaultProperties.putAll(getValues(DEFAULT_GROUP_NAME))
    }
  }

  fun getValue(groupName: String, propertyName: String): String? = configFile[groupName, propertyName]
  fun getValues(groupName: String): Map<String, String?> = configFile[groupName] ?: emptyMap()
  fun setValue(groupName: String, propertyName: String, value: String?) {
    configFile.put(groupName, propertyName, value)
  }

  fun deleteGroup(name: String) {
    if (DEFAULT_GROUP_NAME == name) {
      myDefaultProperties.clear()
    }
    // remove group from groups
    setValue(GROUPS_GROUP_NAME, name, null)
    configFile.remove(name)
  }

  fun addGroup(name: String, patterns: String?, properties: Map<String, String?>) {
    setValue(GROUPS_GROUP_NAME, name, patterns)
    addProperties(name, properties)
  }

  private fun addProperties(groupName: String, properties: Map<String, String?>) {
    for ((key, value) in properties) {
      setValue(groupName, key, value)
    }
  }

  fun modifyGroup(name: String, patterns: String?, delete: Collection<String>, addOrModify: Map<String, String?>, isDefault: Boolean) {
    if (!isDefault) {
      setValue(GROUPS_GROUP_NAME, name, patterns)
    }
    val deletedPrepared = HashMap<String, String?>(delete.size)
    for (property in delete) {
      deletedPrepared.put(property, null)
    }
    addProperties(name, deletedPrepared)
    addProperties(name, addOrModify)
  }

  fun save() = try {
    _configFile.store(myPath.toFile())
    updateGroups(true)
  }
  catch (e: IOException) {
    LOG.info("Could not save $myPath", e)
  }

  companion object {
    @JvmField
    val SERVERS_FILE_NAME = "servers"
    @JvmField
    val CONFIG_FILE_NAME = "config"

    @JvmField
    val DEFAULT_GROUP_NAME = "global"
    @JvmField
    val GROUPS_GROUP_NAME = "groups"

    @JvmStatic
    fun getNewGroupName(host: String, configFile: IdeaSVNConfigFile): String {
      var groupName = host
      val groups = configFile.allGroups
      while (StringUtil.isEmptyOrSpaces(groupName) || groups.containsKey(groupName)) {
        groupName += "1"
      }
      return groupName
    }

    @JvmStatic
    fun getPropertyIdea(host: String, serversFile: Couple<IdeaSVNConfigFile>, name: String): String? {
      val groupName = getGroupName(getValues(serversFile, GROUPS_GROUP_NAME), host)
      if (groupName != null) {
        val hostProps = getValues(serversFile, groupName)
        val value = hostProps[name]
        if (value != null) {
          return value
        }
      }
      return getValues(serversFile, DEFAULT_GROUP_NAME)[name]
    }

    @JvmStatic
    fun checkHostGroup(url: String, patterns: String?, exceptions: String?): Boolean {
      val svnurl: Url
      try {
        svnurl = createUrl(url)
      }
      catch (e: SvnBindException) {
        return false
      }

      val host = svnurl.host
      return matches(patterns, host) && !matches(exceptions, host)
    }

    private fun matches(patterns: String?, host: String) = patterns.orEmpty().split(',').any { matchesPattern(it.trim(), host) }
    private fun matchesPattern(pattern: String, host: String) = try {
      Perl5Matcher().matches(host, GlobCompiler().compile(pattern))
    }
    catch (e: MalformedPatternException) {
      LOG.debug("Could not compile pattern $pattern", e)
      false
    }

    @JvmStatic
    fun getGroupForHost(host: String, serversFile: IdeaSVNConfigFile): String? {
      val groups = serversFile.allGroups
      for ((key, value) in groups) {
        if (matches(value.patterns, host)) return key
      }
      return null
    }

    private fun getGroupName(groups: Map<String, String?>, host: String): String? {
      for ((key, value) in groups) {
        if (matches(value, host)) return key
      }
      return null
    }

    @JvmStatic
    fun isTurned(value: String?) = value == null || "yes".equals(value, true) || "on".equals(value, true) || "true".equals(value, true)

    @JvmStatic
    fun getValue(files: Couple<IdeaSVNConfigFile>, groupName: String, propertyName: String) =
      files.second.getValue(groupName, propertyName) ?: files.first.getValue(groupName, propertyName)

    @JvmStatic
    fun getValues(files: Couple<IdeaSVNConfigFile>, groupName: String): Map<String, String?> =
      union(files.first.getValues(groupName), files.second.getValues(groupName))
  }
}

private class MyIni : Ini() {
  override fun newBuilder() = MyIniBuilder(this)
  override fun store(output: Writer) = store(MyIniFormatter(this, config, output))
}

private class MyIniBuilder(ini: Ini) : IniBuilder() {
  init {
    setIni(ini)
  }

  private var isAfterComment = false;

  override fun startSection(sectionName: String) {
    super.startSection(sectionName)
    isAfterComment = true
  }

  override fun endSection() {
    handleAfterComment()
    isAfterComment = false
    super.endSection()
  }

  override fun handleComment(comment: String) {
    handleHeaderComment()
    handleAfterComment()

    lastComment = (if (!lastComment.isNullOrEmpty()) lastComment + config.lineSeparator else "") + comment
  }

  override fun handleOption(name: String, value: String?) {
    handleAfterComment()
    super.handleOption(name, value)
  }

  private fun handleHeaderComment() {
    if (isHeader && lastComment != null) {
      profile.comment = lastComment
      lastComment = null
      isHeader = false
    }
  }

  private fun handleAfterComment() {
    if (isAfterComment && lastComment != null) {
      (profile as Ini).putMeta("after-comment", currentSection.name, lastComment)
      lastComment = null
      isAfterComment = false
    }
  }
}

private class MyIniFormatter(private val ini: Ini, config: Config, output: Writer) : IniFormatter() {
  init {
    this.config = config
    this.output = output as? PrintWriter ?: PrintWriter(output)
  }

  override fun startSection(sectionName: String) {
    super.startSection(sectionName)
    handleComment(ini.getMeta("after-comment", sectionName) as String?)
  }
}